crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -750,41 +750,58 @@ unsafe impl<'a, T: Resource> ReadOnlySystemParam for Res<'a, T> {}
 // SAFETY: Res ComponentId access is applied to SystemMeta. If this Res
 // conflicts with any prior access, a panic will occur.
 unsafe impl<'a, T: Resource> SystemParam for Res<'a, T> {
-    type State = ComponentId;
+    type State = (ComponentId, ComponentId);
     type Item<'w, 's> = Res<'w, T>;
 
     fn init_state(world: &mut World) -> Self::State {
-        world.components_registrator().register_resource::<T>()
+        (
+            world.components_registrator().register_resource::<T>(),
+            world
+                .components_registrator()
+                .register_component::<IsResource>(),
+        )
     }
 
     fn init_access(
-        &component_id: &Self::State,
+        (component_id, is_resource_id): &Self::State,
         system_meta: &mut SystemMeta,
         component_access_set: &mut FilteredAccessSet,
         _world: &mut World,
     ) {
         let combined_access = component_access_set.combined_access();
-        assert!(
-            !combined_access.has_resource_write(component_id),
+        if combined_access.has_resource_write(*component_id) {
+            panic!(
             "error[B0002]: Res<{}> in system {} conflicts with a previous ResMut<{0}> access. Consider removing the duplicate access. See: https://bevy.org/learn/errors/b0002",
             DebugName::type_name::<T>(),
-            system_meta.name,
-        );
+            system_meta.name);
+        }
 
-        component_access_set.add_unfiltered_resource_read(component_id);
+        let mut filter = FilteredAccess::default();
```

Comment:
I think we need to make a corresponding change to `FilteredResources` and `FilteredResourcesMut` so that they correctly conflict with queries.  

The `has_read/write_all_resources()` branches would probably do `filter.read_all_components(); filter.and_with(*is_resource_id);` to declare access to all components with `IsResource`, while the other branch would use the code you have here in a loop.  

Alternately, it might make sense to modify `add_unfiltered_resource_read/write()` itself so that the code could be shared.  I guess that would also affect `NonSend`, but requesting additional access like that should be fine.